aptos-move/block-executor/src/txn_last_input_output.rs

```diff
@@ -50,58 +53,184 @@ macro_rules! with_success_or_skip_rest {
+        })
+    }
+
+    fn check_success_or_skip_status(&self) -> Result<&O, PanicError> {
+        if self.output_status_kind != OutputStatusKind::Success
+            && self.output_status_kind != OutputStatusKind::SkipRest
+        {
+            return Err(code_invariant_error(format!(
+                "Output status {:?}!= success or skip rest",
+                self.output_status_kind
+            )));
+        }
+
+        Ok(self
+            .output
+            .as_ref()
+            .expect("Output must be set when status is success or skip rest"))
+    }
 }
 
-pub struct TxnLastInputOutput<T: Transaction, O: TransactionOutput<Txn = T>, E: Debug> {
-    inputs: Vec<CachePadded<ArcSwapOption<TxnInput<T>>>>, // txn_idx -> input.
+pub struct TxnLastInputOutput<T: Transaction, O: TransactionOutput<Txn = T>> {
+    inputs: Vec<CachePadded<ArcSwapOption<TxnInput<T>>>>, // txn_idx -> input (read set).
 
-    // TODO: Consider breaking down the outputs when storing (avoid traversals, cache below).
-    outputs: Vec<CachePadded<ArcSwapOption<ExecutionStatus<O, E>>>>, // txn_idx -> output.
+    output_wrappers: Vec<CachePadded<Mutex<OutputWrapper<T, O>>>>,
     // Used to record if the latest incarnation of a txn was a failure due to the
     // speculative nature of parallel execution.
     speculative_failures: Vec<CachePadded<AtomicBool>>,
 }
 
-impl<T: Transaction, O: TransactionOutput<Txn = T>, E: Debug + Send + Clone>
-    TxnLastInputOutput<T, O, E>
-{
+impl<T: Transaction, O: TransactionOutput<Txn = T>> TxnLastInputOutput<T, O> {
     /// num_txns passed here is typically larger than the number of txns in the block,
     /// currently by 1 to account for the block epilogue txn.
     pub fn new(num_txns: TxnIndex) -> Self {
         Self {
             inputs: (0..num_txns)
                 .map(|_| CachePadded::new(ArcSwapOption::empty()))
                 .collect(),
-            outputs: (0..num_txns)
-                .map(|_| CachePadded::new(ArcSwapOption::empty()))
+            output_wrappers: (0..num_txns)
+                .map(|_| {
+                    CachePadded::new(Mutex::new(OutputWrapper::empty_with_status(
+                        OutputStatusKind::None,
+                    )))
+                })
                 .collect(),
             speculative_failures: (0..num_txns)
                 .map(|_| CachePadded::new(AtomicBool::new(false)))
                 .collect(),
         }
     }
 
-    pub(crate) fn record(
+    pub(crate) fn record<E: Debug>(
         &self,
         txn_idx: TxnIndex,
         input: TxnInput<T>,
         output: ExecutionStatus<O, E>,
-    ) {
+        block_gas_limit_type: &BlockGasLimitType,
+        user_txn_bytes_len: u64,
+    ) -> Result<(), PanicError> {
         self.speculative_failures[txn_idx as usize].store(false, Ordering::Relaxed);
+        *self.output_wrappers[txn_idx as usize].lock() = OutputWrapper::from_execution_status(
```

Comment:
is getting rid of the Arc on the output intended?

Reply:
yes! not needed